File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.05.PizzaCalories/Topping.php
<?php


class Topping
{
    const CALORIES_PER_GRAM = 2;
    const TOPPING_TYPES =
        [
            "meat" => 1.2,
            "veggies" => 0.8,
            "cheese" => 1.1,
            "sauce" => 0.9
        ];
    /**
     * @var string
     */
    private $type;
    /**
     * @var int
     */
    private $weight;

    /**
     * Topping constructor.
     * @param string $type
     * @param int $weight
     * @throws Exception
     */
    public function __construct(string $type, int $weight)
    {
        $this->setType($type);
        $this->setWeight($weight);
    }

    /**
     * @param string $type
     * @throws Exception
     */
    protected function setType(string $type): void
    {
        if (!array_key_exists($type, self::TOPPING_TYPES)) {
            $type[0] = strtoupper($type[0]);
            throw new Exception("Cannot place {$type} on top of your pizza.");
        }
        $this->type = $type;
    }

    /**
     * @param int $weight
     * @throws Exception
     */
    protected function setWeight(int $weight): void
    {
        if ($weight < 1 || $weight > 50) {
            $this->type[0] = strtoupper($this->type[0]);
            throw new Exception("{$this->type} weight should be in the range [1..50].");

        }
        $this->weight = $weight;
    }

    public function getCalories()
    {
        return $this->weight
            * self::CALORIES_PER_GRAM
            *self::TOPPING_TYPES[$this->type];
    }
}

File: Courses/C# OOP/Basics/08. Encapsulation Exercise/08. Encapsulation Exercise/05.PizzaCalories/Topping.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace PizzaCalories
{
    public class Topping
    {
        private string type;
        private double weight;

        public Topping(string type, double weight)
        {
            Type = type;
            Weight = weight;
        }

        public string Type
        {
            get => type;
            set
            {
                switch (value.ToLower())
                {
                    case "meat":
                    case "veggies":
                    case "cheese":
                    case "sauce":
                        type = value.ToLower();
                        break;
                    default:
                        throw new ArgumentException($"Cannot place {value} on top of your pizza.");
                }
            }
        }
        public double Weight
        {
            get => weight; 
            set
            {
                if (value < 1 || value > 50)
                {
                    var t = Type[0].ToString().ToUpper() + Type.Substring(1);
                    throw new ArgumentException($"{t} weight should be in the range [1..50].");
                }
                weight = value;
            }
        }

        public double ToppingCalories { get => this.CalculateToppingCalories(); }

        public double CalculateToppingCalories()
        {
            double topppingModifier = Type == "meat" ? 1.2 : Type == "veggies" ? 0.8 : Type == "cheese"? 1.1: 0.9;
            return 2 * Weight * topppingModifier;
        }
    }
}
